### PR TITLE
Created new Bridge method: getEstimatedFeesForNextPegOutEvent

### PR DESCRIPTION
**Pull request by kelvinator07**

### CODE REVIEW

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java [hunk 1]
```diff
@@ -7609,6 +7609,40 @@ public void getQueuedPegoutsCount_after_RSKIP271_activation() throws IOException
...
+                .withActivations(activations)
+                .withProvider(provider)
+                .build();
+
+        verify(provider, never()).getFeePerKb();
+        assertEquals(Coin.ZERO, bridgeSupport.getEstimatedFeesForNextPegOutEvent());
+    }
+
+    @Test
+    public void getEstimatedFeesForNextPegOutEvent_after_RSKIP271_activation() throws IOException {
```
Comment (josedahlquist) [inline rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java:7629]: Please add additional tests:
- rskip is activated but there are no pegout requests, should return a 0.
- fees calculated with and without an erp Federation
Some other cases you may think of @crove-iov or @marcos-iov ?

File: rskj-core/src/test/java/co/rsk/peg/BridgeUtilsTest.java [hunk 2]
```diff
@@ -2093,16 +2093,16 @@ public void extractHash160FromBytes() throws BridgeIllegalArgumentException {
     @Test
     public void calculatePegoutTxSize() {
```
Comment (josedahlquist) [inline rskj-core/src/test/java/co/rsk/peg/BridgeUtilsTest.java:2094]: we probably need more cases here as well:
- Federation of different sizes (and ERP/regular)
- what happens if no inputs or outputs are sent?
- exercising some limits (100 inputs and outputs)

File: rskj-core/src/main/java/co/rsk/peg/BridgeMethods.java [hunk 3]
```diff
@@ -679,6 +679,17 @@
             (BridgeMethodExecutorTyped) Bridge::getQueuedPegoutsCount,
             activations -> activations.isActive(RSKIP271),
             fixedPermission(false)
+    ),
+    GET_ESTIMATED_FEES_FOR_NEXT_PEGOUT_EVENT(
+            CallTransaction.Function.fromSignature(
+                    "getEstimatedFeesForNextPegOutEvent",
+                    new String[]{},
+                    new String[]{"int256"}
```
[the remark this answers is not in the document]
Comment (kelvinator07) [reply rskj-core/src/main/java/co/rsk/peg/BridgeMethods.java]: Done

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java [hunk 4]
```diff
@@ -2511,6 +2511,30 @@ public int getQueuedPegoutsCount() throws IOException {
...
+
+        final int INPUT_MULTIPLIER = 2;
+
+        int pegoutRequestsCount = getQueuedPegoutsCount();
+
+        if (!activations.isActive(RSKIP271) || pegoutRequestsCount == 0) {
+            return Coin.ZERO;
+        }
+
+        int totalPegouts = pegoutRequestsCount + 2; // N + 2 outputs
```
[the remark this answers is not in the document]
Comment (kelvinator07) [reply rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java]: Done

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java [hunk 5]
```diff
@@ -2511,6 +2511,30 @@ public int getQueuedPegoutsCount() throws IOException {
...
+
+        int pegoutTxSize = BridgeUtils.calculatePegoutTxSize(getActiveFederation(), INPUT_MULTIPLIER, totalPegouts);
+
+        Coin feePerKB = getFeePerKb();
+
+        return feePerKB
+                .multiply(pegoutTxSize) // times the size in bytes
+                .multiply(totalPegouts) // times the total Pegout count
```
[the remark this answers is not in the document]
Comment (kelvinator07) [reply rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java]: Thanks For The Catch. Resolved

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java [hunk 6]
```diff
@@ -7609,6 +7609,40 @@ public void getQueuedPegoutsCount_after_RSKIP271_activation() throws IOException
...
+        when(provider.getReleaseRequestQueueSize()).thenReturn(3);
+        when(provider.getFeePerKb()).thenReturn(Coin.MILLICOIN);
+
+        BridgeSupport bridgeSupport = bridgeSupportBuilder
+                .withProvider(provider)
+                .withActivations(activations)
+                .withBridgeConstants(bridgeConstants)
+                .build();
+
+        assertEquals(Coin.valueOf(75900), bridgeSupport.getEstimatedFeesForNextPegOutEvent());
```
Comment (marcos-iov) [inline rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java:7643]: I don't like the idea of asserting against a magic value of `75900`. How was that calculated? It would be better if the test does the same calculation, taking into account release request queue size and fee per kb values

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java (hunk 1 above)
Comment (crove-iov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java:7629]: * what happens if fees are below dust values?
* limit case of only 1 pegout in queue should pay all fees.

**Review by crove-iov — COMMENTED** [review]
I added a couple of test cases, we can discuss further if required.

File: rskj-core/src/test/java/co/rsk/peg/BridgeUtilsTest.java [hunk 7]
```diff
@@ -2093,16 +2093,16 @@ public void extractHash160FromBytes() throws BridgeIllegalArgumentException {
...
             0,
             networkParameters
         );
 
-        int pegoutTxSize = BridgeUtils.calculatePegoutTxSize(fed, 1, 1, 1, 1);
+        int pegoutTxSize = BridgeUtils.calculatePegoutTxSize(fed, 1, 1);
 
-        // 58 accounts for all the added values as part of a peg-out tx
-        int calculation = fed.getRedeemScript().getProgram().length+ 58;
+        // 152 accounts for all the added values as part of a peg-out tx
```
Comment (marcos-iov) [inline rskj-core/src/test/java/co/rsk/peg/BridgeUtilsTest.java:2104]: Why did this change from 58 to 152?

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java (hunk 1 above)
Comment (marcos-iov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java:7629]: Can't think of any right now. Maybe the with/without ERP and other tests should be done for `calculatePegoutTxSize` method.

Adding a test for the release requests queue being empty should be enough to cover this method I think

File: rskj-core/src/test/java/co/rsk/peg/BridgeTest.java [hunk 8]
```diff
@@ -813,6 +813,36 @@ public void getQueuedPegoutsCount_after_RSKIP271_activation() throws VMException
...
+
+        when(bridgeSupportMock.getEstimatedFeesForNextPegOutEvent()).thenReturn(Coin.SATOSHI);
+
+        CallTransaction.Function function = BridgeMethods.GET_ESTIMATED_FEES_FOR_NEXT_PEGOUT_EVENT.getFunction();
+        byte[] data = function.encode(new Object[]{ });
+        byte[] result = bridge.execute(data);
+
+        assertEquals(1, ((BigInteger)function.decodeResult(result)[0]).intValue());
+        // Also test the method itself
+        assertEquals(1, bridge.getEstimatedFeesForNextPegOutEvent(new Object[]{ }));
```
[the remark this answers is not in the document]
Comment (kelvinator07) [reply rskj-core/src/test/java/co/rsk/peg/BridgeTest.java]: Done

File: rskj-core/src/test/java/co/rsk/peg/BridgeTest.java [hunk 9]
```diff
@@ -813,6 +813,36 @@ public void getQueuedPegoutsCount_after_RSKIP271_activation() throws VMException
...
+        BridgeSupport bridgeSupportMock = mock(BridgeSupport.class);
+        Bridge bridge = getBridgeInstance(bridgeSupportMock);
+
+        when(bridgeSupportMock.getEstimatedFeesForNextPegOutEvent()).thenReturn(Coin.SATOSHI);
+
+        CallTransaction.Function function = BridgeMethods.GET_ESTIMATED_FEES_FOR_NEXT_PEGOUT_EVENT.getFunction();
+        byte[] data = function.encode(new Object[]{ });
+        byte[] result = bridge.execute(data);
+
+        assertEquals(1, ((BigInteger)function.decodeResult(result)[0]).intValue());
```
[the remark this answers is not in the document]
Comment (kelvinator07) [reply rskj-core/src/test/java/co/rsk/peg/BridgeTest.java]: Done

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java (hunk 6 above)
Comment (kelvinator07) [reply rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java]: Resolved

File: rskj-core/src/main/java/co/rsk/peg/BridgeUtils.java [hunk 10]
```diff
@@ -563,38 +563,35 @@ public static int getRegularPegoutTxSize(@Nonnull Federation federation) {
...
             federation,
             INPUT_MULTIPLIER,
-            SIGNATURE_MULTIPLIER,
-            OUTPUT_MULTIPLIER,
-            OUTPUT_SIZE
+            OUTPUT_MULTIPLIER
         );
     }
 
     public static int calculatePegoutTxSize(
```
[the remark this answers is not in the document]
Comment (kelvinator07) [reply rskj-core/src/main/java/co/rsk/peg/BridgeUtils.java]: Done

File: rskj-core/src/test/java/co/rsk/peg/BridgeUtilsTest.java (hunk 7 above)
Comment (kelvinator07) [reply rskj-core/src/test/java/co/rsk/peg/BridgeUtilsTest.java]: Resolved

File: rskj-core/src/test/java/co/rsk/peg/BridgeUtilsLegacyTest.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+                FederationMember.getFederationMembersFromKeys(keys),
+                Instant.now(),
+                0,
+                networkParameters
+        );
+
+        int pegoutTxSize = BridgeUtilsLegacy.calculatePegoutTxSize(activations, federation, 2, 2);
+
+        // The difference between the calculated size and a real tx size should be smaller than 1% in any direction
+        int origTxSize = 2076; // Data for 2 inputs, 2 outputs Based on Pegouts From Blockchain.info Explorer
```
Comment (marcos-iov) [inline rskj-core/src/test/java/co/rsk/peg/BridgeUtilsLegacyTest.java:48]: Maybe include a link to the transaction in a explorer, for future reference

File: rskj-core/src/test/java/co/rsk/peg/BridgeUtilsTest.java [hunk 12]
```diff
@@ -2092,22 +2092,32 @@ public void extractHash160FromBytes() throws BridgeIllegalArgumentException {
...
+        // The difference between the calculated size and a real tx size should be smaller than 1% in any direction
+        int origTxSize = 2076; // Data for 2 inputs, 2 outputs Based on Pegouts From Blockchain.info Explorer
+        int difference = origTxSize - pegoutTxSize;
+        double tolerance = origTxSize * .01;
 
-        // 58 accounts for all the added values as part of a peg-out tx
-        int calculation = fed.getRedeemScript().getProgram().length+ 58;
-        assertEquals(calculation, pegoutTxSize);
+        assertTrue(difference < tolerance && difference > -tolerance);
     }
```
Comment (marcos-iov) [inline rskj-core/src/test/java/co/rsk/peg/BridgeUtilsTest.java:2114]: Let's add tests for other transactions sizes. 

- 3 inputs and 2 outputs
- 4 inputs, etc
- Also multiple outputs since we expect to start batching pegouts

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java [hunk 13]
```diff
@@ -7609,6 +7609,134 @@ public void getQueuedPegoutsCount_after_RSKIP271_activation() throws IOException
...
+                mock(BridgeEventLogger.class),
+                new BtcLockSenderProvider(),
+                new PeginInstructionsProvider(),
+                mock(Repository.class),
+                mock(Block.class),
+                mock(Context.class),
+                federationSupport,
+                mock(BtcBlockStoreWithCache.Factory.class),
+                activations
+        );
```
Comment (marcos-iov) [inline rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java:7655]: You can simply use BridgeSupportBuilder class here

Comment (kelvinator07) [reply rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java:7655]: Using BridgeSupportBuilder Doesn't Instantiate Federation Support For Some Weird Reasons

Comment (kelvinator07) [reply rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java]: Resolved

File: rskj-core/src/test/java/co/rsk/peg/BridgeUtilsTest.java (hunk 12 above)
Comment (kelvinator07) [reply rskj-core/src/test/java/co/rsk/peg/BridgeUtilsTest.java:2141]: Done

File: rskj-core/src/test/java/co/rsk/peg/BridgeUtilsLegacyTest.java (hunk 11 above)
Comment (kelvinator07) [reply rskj-core/src/test/java/co/rsk/peg/BridgeUtilsLegacyTest.java]: Done

File: rskj-core/src/main/java/co/rsk/peg/BridgeUtilsLegacy.java [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+                                               Federation federation,
+                                               int inputs,
+                                               int outputs) {
+        if (activations.isActive(ConsensusRule.RSKIP271)) {
+            throw new DeprecatedMethodCallException(
+                    "Calling BridgeUtilsLegacy.calculatePegoutTxSize method after RSKIP271 activation"
+            );
+        }
+
+        final int SIGNATURE_MULTIPLIER = 72;
```
[the remark this answers is not in the document]
Comment (kelvinator07) [reply rskj-core/src/main/java/co/rsk/peg/BridgeUtilsLegacy.java]: Resolved

File: rskj-core/src/test/java/co/rsk/peg/BridgeUtilsTest.java (hunk 2 above)
Comment (kelvinator07) [reply rskj-core/src/test/java/co/rsk/peg/BridgeUtilsTest.java]: Done

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java (hunk 1 above)
Comment (kelvinator07) [reply rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java:7629]: Resolved

File: rskj-core/src/test/java/co/rsk/peg/BridgeUtilsTest.java [hunk 15]
```diff
@@ -2381,6 +2584,33 @@ private BtcTransaction createPegOutTx(
         return btcTx;
     }
 
+    private BtcTransaction createPegOutTx(int inputSize, int outputSize, Federation federation, List<BtcECKey> keys) {
```
[the remark this answers is not in the document]
Comment (kelvinator07) [reply rskj-core/src/test/java/co/rsk/peg/BridgeUtilsTest.java:2587]: This Method Calls A non-static method signWithNecessaryKeys, Which Might Be a Challenge

File: rskj-core/src/test/java/co/rsk/peg/BridgeUtilsTest.java [hunk 16]
```diff
@@ -2194,6 +2389,14 @@ public void scriptCorrectlySpends_invalidScript() {
         assertFalse(BridgeUtils.scriptCorrectlySpendsTx(tx, 0, genesisFederation.getP2SHScript()));
     }
 
+    private List<BtcECKey> createBtcECKeys(int keysCount) {
```
[the remark this answers is not in the document]
Comment (kelvinator07) [reply rskj-core/src/test/java/co/rsk/peg/BridgeUtilsTest.java]: Done

File: rskj-core/src/test/java/co/rsk/peg/BridgeUtilsTest.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (marcos-iov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeUtilsTest.java:2579]: Ok, probably not worth moving all those methods right now. Thanks

File: rskj-core/src/test/java/co/rsk/peg/BridgeUtilsLegacyTest.java [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+        BridgeUtilsLegacy.calculatePegoutTxSize(activations, federation, 0, 0);
+    }
+
+    private List<BtcECKey> createBtcECKeys(int keysCount) {
+        List<BtcECKey> keys = new ArrayList<>();
+        for (int i = 0; i < keysCount; i++) {
+            keys.add(new BtcECKey());
+        }
+        return keys;
+    }
```
[the remark this answers is not in the document]
Comment (kelvinator07) [reply rskj-core/src/test/java/co/rsk/peg/BridgeUtilsLegacyTest.java]: Done

**Review by marcos-iov — APPROVED** [review]
Well done!

**Comment by kelvinator07** [comment]
> Well done!

Gracias Marcos!

**Review by josedahlquist — APPROVED** [review]
